src/rail_cmd.cpp

```diff
@@ -1545,6 +1545,81 @@ static Vehicle *UpdateTrainPowerProc(Vehicle *v, void *data)
 	return nullptr;
 }
 
+struct EnsureNoIncompatibleRailtypeTrainOnGroundData {
+	int z;
+	RailType type;
+};
+
+static Vehicle *EnsureNoIncompatibleRailtypeTrainProc(Vehicle *v, void *data)
+{
+	if (v->type != VEH_TRAIN) {
+		if (v->type == VEH_DISASTER || v->type == VEH_AIRCRAFT) return nullptr; // Don't care about flying things
+		return v; // Non-trains are always in the way
+	}
+
+	const EnsureNoIncompatibleRailtypeTrainOnGroundData *procdata = (EnsureNoIncompatibleRailtypeTrainOnGroundData *)data;
+
+	if (v->z_pos > procdata->z) return nullptr;
+	Train *t = Train::From(v)->First();
+	if (HasBit(t->compatible_railtypes, procdata->type) && HasPowerOnRail(t->railtype, procdata->type)) return nullptr;
```

Comment:
`HasPowerOnRail(t->railtype, procdata->type)` is not correct as this only tests the rail type of the first engine.
It's perfectly legal to create mixed power consists of compatible but not necessarily mutually powered rail types.

See the calculation of Train::compatible_railtypes in Train::ConsistChanged.

For testing you also need to note the current value of VRF_EL_ENGINE_ALLOWED_NORMAL_RAIL

Reply:
Yes, I did wonder. Not as simple as I thought! I'll get there... probably.